### ISSUE TITLE
Add lo and hi parameters to bisect methods

### ISSUE THREAD

**Issue by Evidlo** [grantjenks/python-sortedcontainers#115]
I'd like to propose a `recursive` kwarg to `SortedDict.__init__` so that the following

``` python
SortedDict({
    1:SortedDict({
        1:'hello',
        2:'world'
    }),
    2:SortedDict({
        1:'Hello',
        2:'World
    })
})
```

could be shortened to

``` python
SortedDict({
    1:{
        1:'hello',
        2:'world'
    },
    2:{
        1:'Hello',
        2:'World
    }
}, recursive=True)
```

This might also apply to `SortedList` and `SortedSet`.

You'd have to decide what happens with something like `{1:[{1:'foo'}], 2:[{1:'bar'}]}` and how deep to go.  Simplest option would be to stop recursing when you encounter a non-dict.

**Comment by grantjenks** [comment]
Interesting. Not sure I’ve ever had a use case. What’s yours?

**Comment by Evidlo** [comment]
My use case is looking up 

``` python
regions =  SortedDict({
    500: SortedDict({
        500: 'region A',
        1000: 'region B'
    }),
    500: SortedDict({
        1000 // 2: 'region C',
        1000: 'region D'
    })
})

def lookup(x, y, regions):
    regions = self.regions()
    row_index = regions.keys()[regions.bisect(y)]
    row = regions[row_index]
    col_index = row.keys()[row.bisect(x)]
    return row[col_index]

lookup(250, 250, regions)
# 'region A'
```

It might be nice to have a shorthand way to do `regions[regions.keys()[regions.bisect(y)]]`.  Maybe `regions.get_bisect(y)`?

On a related note, if `x` or `y` are greater than 1000.  In the bisect module, I can use the `hi` and `low` kwargs to ensure `regions.bisect(y)` will always return a valid key index.  Is there anything like this for SortedDict?

**Comment by grantjenks** [comment]
Your coordinate system is a bit unusual to me. So region A is like (0-500, 0-500) and region B is (0-500, 500-1000) and region C is (500-1000, 0-500) and region D is (500-1000, 500-1000). And you want to identify what region a point belongs to?

I think the `lo` and `hi` parameters could be a good addition for `bisect`. The SortedList class uses the `bisect` module so I think it should a straightforward add. Pull request welcome.

**Comment by grantjenks** [comment]
I remembered last night that there’s a reason for not supporting lo and hi. Indexing is relatively expensive compared to bisecting. So unless the length of sorted lists is very high (I think billions), it doesn’t help much.

The bisect methods are mostly for compatibility with the old blist module. The irange method should usually be faster if you just want the nearest key. The bintrees module had something like that. I resisted copying their API initially but your use case and others is making it seem more attractive.
